File: src/leetCode/java/Solution588.java
package leetCode.java;

import java.util.*;

//FileSystem(), ls(), addContentToFile(), readContentFromFile(): O(1) O(n), mkdir(): O(m) O(n)
//n = 檔案或資料夾數量, m = path.length()
class FileSystem {//Solution588 $
    public Map<String, Set<String>> dirs;
    public Map<String, String> files;

    public FileSystem() {
        dirs = new HashMap<>();
        files = new HashMap<>();
        dirs.put("/", new HashSet<>());
    }

    public List<String> ls(String path) {
        if(files.containsKey(path)){
            int i = path.lastIndexOf("/");
            List<String> res = new ArrayList<>();
            res.add(path.substring(i + 1));
            return res;
        }else if(dirs.containsKey(path)){
            return new ArrayList<>(dirs.get(path));
        }else{
            return new ArrayList<>();
        }
    }

    public void mkdir(String path) {
        if("/".equals(path)) return;

        String[] p = path.split("/");
        String dir = "/";
        for(int i=1; i<p.length; i++){//掉過第一個'/'
            if(!dirs.containsKey(dir))
                dirs.put(dir, new HashSet<>());

            dirs.get(dir).add(p[i]);
            dir += ("/".equals(dir) ? p[i] : "/" + p[i]);
        }
        if(!dirs.containsKey(dir))
            dirs.put(dir, new HashSet<>());
    }

    public void addContentToFile(String filePath, String content) {
        int i = filePath.lastIndexOf("/");
        String dir = filePath.substring(0, i);
        String fileName = filePath.substring(i + 1);
        if(dir.isEmpty()) dir = "/";

        if(!dirs.containsKey(dir)) mkdir(dir);
        dirs.get(dir).add(fileName);
        files.put(filePath, content);
    }

    public String readContentFromFile(String filePath) {
        if(!files.containsKey(filePath)) return "";
        return files.get(filePath);
    }
}

File: c++/Solution588.cpp
//FileSystem(), ls(), addContentToFile(), readContentFromFile(): O(1) O(n), mkdir(): O(m) O(n)
//n = 檔案或資料夾數量, m = path.length()
class FileSystem {//Solution588 $
public:
    unordered_map<string, set<string>> dirs;//<path, 檔名>
    unordered_map<string, string> files;//<檔名, 內容>

    FileSystem() {
        dirs["/"];
    }

    vector<string> ls(string path) {
        if(files.count(path)){//路徑是檔案時
            int i = path.find_last_of("/");//等同java lastIndexOf()
            return {path.substr(i + 1)};
        }

        set<string> names = dirs[path];
        return vector<string>(names.begin(), names.end());
    }

    void mkdir(string path) {
        if(path == "/") return;

        string dir = "/", folder = "";
        for(int i=1; i<path.length(); ++i){//略過第一個'/'
            if(path[i] == '/'){
                dirs[dir].insert(folder);
                dir += (dir == "/" ? folder : "/" + folder);
                folder.clear();
            }else{
                folder.push_back(path[i]);
            }
        }
        dirs[dir].insert(folder);
        dir += (dir == "/" ? folder : "/" + folder);
        dirs[dir];
    }

    void addContentToFile(string filePath, string content) {
        int i = filePath.find_last_of("/");
        string dir = filePath.substr(0, i);
        string fileName = filePath.substr(i + 1);
        if(dir.empty()) dir = "/";//避免filePath="/a"

        if(!dirs.count(dir)) mkdir(dir);
        dirs[dir].insert(fileName);
        files[filePath].append(content);
    }

    string readContentFromFile(string filePath) {
        return files[filePath];
    }

};